src/main/java/rockset/RocksetSinkTask.java

```diff
@@ -71,164 +54,98 @@ public void start(Map<String, String> settings) {
   @Override
   public void open(Collection<TopicPartition> partitions) {
-    log.debug(String.format("Opening %d partitions: %s", partitions.size(), partitions));
-    partitions.forEach(
-        tp -> {
-          Preconditions.checkState(!futureMap.containsKey(tp));
-          futureMap.put(tp, new ArrayList<>());
-        });
+    log.debug("Opening {} partitions: {}", partitions.size(), partitions);
   }
 
   @Override
   public void close(Collection<TopicPartition> partitions) {
-    log.debug(String.format("Closing %d partitions: %s", partitions.size(), partitions));
-    partitions.forEach(
-        tp -> {
-          Preconditions.checkState(futureMap.containsKey(tp));
-          futureMap.remove(tp);
-        });
+    log.debug("Closing {} partitions: {}", partitions.size(), partitions);
   }
 
-  // put() doesn't need to block until the writes complete, that is what flush() is for
   @Override
   public void put(Collection<SinkRecord> records) {
-    if (records.size() == 0) {
-      log.debug("zero records in put call, returning");
-      return;
-    }
-
-    groupRecordsByTopicPartition(records)
-        .forEach(
-            (tp, recordBatch) -> {
-              try {
-                RetriableTask task =
-                    new RetriableTask(
-                        taskExecutorService,
-                        retryExecutorService,
-                        () -> addDocs(tp.topic(), recordBatch));
-
-                // this should only block if all the threads are busy
-                taskExecutorService.submit(task);
-
-                Preconditions.checkState(futureMap.containsKey(tp));
-                futureMap.get(tp).add(task);
-              } catch (InterruptedException e) {
-                throw new ConnectException("Failed to put records", e);
+    Map<TopicPartition, Future<?>> futures =
+        records.stream()
+            .collect(Collectors.groupingBy(r -> new TopicPartition(r.topic(), r.kafkaPartition())))
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    e ->
+                        executorService.submit(
+                            () ->
+                                requestWrapper.addDoc(
+                                    e.getKey().topic(),
+                                    e.getValue(),
+                                    recordParser,
+                                    config.getRocksetBatchSize()))));
+    futures.forEach(
+        (tp, f) -> {
+          try {
+            f.get();
+          } catch (Exception e) {
+            // Stop all tasks still executing since this put() will be retried anyway
+            futures.values().forEach(g -> g.cancel(true));
+            if (isRetriableException(e)) {
+              if (context != null) {
+                context.timeout(config.getRetryBackoffMs());
```

Comment:
should this be set once during initialization? I guess changing this config requires restart anyways?

Reply:
This has to be set every time since Kafka Connect resets it (otherwise you would just retry forever from this point onwards) https://github.com/apache/kafka/blob/9b468fb278701be836a2641650356907bf84860a/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L330-L334

Reply:
ah got it - thanks for the code pointer!